binderhub/builder.py

```diff
@@ -424,9 +424,14 @@ def escape(s):
 
     async def launch(self, kube, provider):
         """Ask JupyterHub to launch the image."""
+        # Load the spec-specific configuration if it has been overridden
+        spec_config = provider.spec_configuration_override
+
         # check quota first
         if provider.has_higher_quota():
```

Comment:
Should we retire this at the same time as we introduce the new functionality?

Reply:
Probably good to have a deprecation transition, in which case include a warning if the old way is used.

Reply:
I'm fine w/ a transition (though we've still only ever had one release of BinderHub lol)

Reply:
True on paper and in practice we release several times a week and people actually use BinderHub so I think we should be nice to people. The "higher quota" thing has only existed for a week or two so :-/

```python
warnings.warn(
    "XXX is deprecated, use YYY instead",
    DeprecationWarning
)
```
is the generic Python warnings thing. It isn't ideal (most people won't see deprecation warnings except in their tests). Does traitlets have a tool for deprecations?

Reply:
Due to the invisible-by-default deprecation warnings, I think that application-level deprecation warnings should actually *not* use DeprecationWarning, so they are visible by default. It's a tricky choice since deprecation warnings are always safely ignored "for now" but if folks never see them, they may as well not exist. As for a transition, I was mostly thinking of ourselves at mybinder.org, since it would allow us to decouple "adopt the new way" from "bump binderhub" PRs without config changes.

FWIW, I don't think the deprecation needs to be a blocker here, as the old way still works. Cleanly deprecating the old way can be a follow-up PR.

> Does traitlets have a tool for deprecations?

Traitlets doesn't have any helpers for deprecations, but the way traitlets works can make writing deprecations handy since deprecations can go in observers, and completely removed from the bodies of methods, etc. for example:

```python
@observe('old_way')
def _old_way_changed(self, change):
    warnings.warn("old way is deprecated as of...new way is...")
    self.new_way = modified_form(change.new)
    # never need to look directly at self.old_way after this

...

def method(self):
    only_use(self.new_way)
```

The simplest version of this can lead to undefined behavior (which wins?) if both the old and new way are specified in the same config, but I think that's not a big deal if the old-way warning is visible.

Reply:
I think using the traitlets `observe` together with printing a warning (via `warnings.warn()`) is a good way to do this.